### ISSUE TITLE
Add compatibility for python3?

### ISSUE THREAD

**Issue by hughperkins** [mattheww/gomill#1]
Hi Matthew,

Just on the off-chance, fancy adding compatibility for python3?  You can use things like `from __future__ import print_function, unicode_literals`, so that it will work with both python2 and python3 :-)

Hugh

**Comment by mattheww** [comment]
That would be a good thing, but I don't think I'm likely to have any time to spend on gomill in the near future. I'd be happy if someone else wanted to, though.

At present gomill uses 8-bit utf-8 strings internally, rather than Python 2 unicode objects, pretty much everywhere. The current code and documentation is fairly careful about the distinction.

Both GTP and SGF are inherently based on 8-bit strings, so switching gomill to generally use Python 3 strings internally might not be the easiest way to go.

**Comment by traveller42** [comment]
I'm currently working through getting this working with Python 3.

I'm starting with the python3 branch @cjlarose created in his fork.  A pass of conversion was completed and all tests were successful using Python 2.

I'm now working through resolving errors running the tests on Python 3.  I may need to do this in two passes.  First, force everything to 8-bit binary types and literals to get it working.  Then, look at points that can be converted properly to unicode.  

The Python core developers suggest using unicode internally and converting to the required encoding as needed at the boundary.  This may not make sense for the reasons mentioned above.

**Comment by cjlarose** [comment]
> I'm starting with the python3 branch @cjlarose created in his fork. A pass of conversion was completed and all tests were successful using Python 2.

Thanks for picking this up! One thing I ran into was deciding whether or not it was worth keeping compatibility with Python 2. Not sure how many people depend on Python 2 support, so it might make sense to aim for just Python 3 support.

> The Python core developers suggest using unicode internally and converting to the required encoding as needed at the boundary. This may not make sense for the reasons mentioned above.

Looks like the latest [SGF spec][1] says most text should be interpreted as ISO-8859-1 unless otherwise specified (the gomill SGF serializer [already does this][3] appropriately). The [GTP spec][2] specifies that only US-ASCII should be used. Both should be pretty easy to convert to `unicode` types internally. I haven't worked closely with this code in a long time, but I suspect you might be better off dealing with Unicode internally as recommended.

[1]: http://www.red-bean.com/sgf/changes.html
[2]: http://www.lysator.liu.se/~gunnar/gtp/gtp2-spec-draft2/gtp2-spec.html#SECTION00030000000000000000
[3]: https://github.com/mattheww/gomill/blob/6aa24d197723c6ce19fb88a6c6c57f7b0b0f87d5/gomill/sgf.py#L587-L588

**Comment by mattheww** [comment]
Note that gomill documents that it accepts non-ascii content in some GTP responses and treats it as utf-8:

> GTP engines may return UTF-8 characters in in response to name, version, gomill-describe_engine, or gomill-explain_last_move.

**Comment by traveller42** [comment]
Thanks for the replies.

 I decided I didn't' gain enough by dropping Python 2 support.  Having the quite extensive test suite available makes this much easier.  I feel like the effort is on the Python 3 side and I get the Python 2.7 support for free.  My plan is to get it working on my system with 2.7 and 3.4.  Once that is stable and correct, I will back test 2.6 and 2.5 to see where it breaks.  I expect to break 2.5,  and I don't intend to fix it if that happens.  Support for 2.6 will depend on the effort (and ugliness) required.  Neither are supported officially by the code developers.  I will also back test 3.3 and 3.2.  Some of the current techniques I'm using may not work with 3.2.  It's support ends early next year, so fixing any issue will again depend on the effort required.  Now that 3.5 has been released, I will verify it before I consider this task complete.

Handling the GTP and SGF specs properly (or at least reasonably), is another kettle of fish.  I'm paying attention to the notes in the code regarding the expected value and the return values.  I've already found something that looks like a bug in the manner Python 3 is passing return values.  I'm going to trust the test cases for now.  After I have something stable, I will go back through and ensure that each boundary touchpoint is using the expected encoding on the external side.  I also want to be sure that data that goes through any of the interfaces round-trip comes out the same.  Where I cannot make that happen, I will ensure the conditions are captured in the documents.

I started with something around 360 errors in Python 3.  I'm under 300 at last check.
